src/classes/BDI_DataImportBatchStatus_TDTM.cls

```diff
@@ -35,193 +35,276 @@
-            openStatusCount = 0;
-            importedStatusCount = 0;
-            failedStatusCount = 0;
+     * A class for determining the status of a single Data Import Batch
+     */
+    private class BatchDataImportStatus {
+        Boolean hasOpenRow = false;
+        Boolean hasImportedRow = false;
+        Boolean hasFailedRow = false;
+
+        /**
+         * @param batch A data import batch with a single row sub-query on the NPSP_Data_Imports__r relationship
+         */
+        void readQueryResult(DataImportBatch__c batch) {
+            if (batch.NPSP_Data_Imports__r != null) {
+                for (DataImport__c dataImport : batch.NPSP_Data_Imports__r) {
+                    readRowStatus(dataImport);
+                }
+            }
         }
 
         /**
-         * @description update the status count based on the incoming di status
-        * @param String status
-        */
-        void updateCounts(String status) {
-            if (String.isBlank(status) || status == BDI_DataImport_API.bdiDryRunValidated
-                                    || status == BDI_DataImport_API.bdiDryRunError) {
-                openStatusCount++;
-            } else if (status == BDI_DataImport_API.bdiImported) {
-                importedStatusCount++;
-            } else if (status == BDI_DataImport_API.bdiFailed) {
-                failedStatusCount++;
+         * Read the status of a single data import row from the inner query on the batch,
+         * and set one of the batch status flags to indicate that the specified type of row was found.
+         * @param dataImport The data import row to get the status of.
+         */
+        void readRowStatus(DataImport__c dataImport) {
+            String status = dataImport.Status__c;
+
+            if (isOpen(status)) {
+                this.hasOpenRow = true;
+                return;
+            }
+
+            if (isImported(status)) {
+                this.hasImportedRow = true;
+                return;
+            }
+
+            if (isFailed(status)) {
+                this.hasFailedRow = true;
+                return;
+            }
+        }
+
+        /**
+         * After ingesting all of the Data Import Batch rows, determine the final batch status.
+         * @return {String} One of 'Failed - Needs Review', 'Open', or 'Completed'
+         */
+        String getBatchStatus() {
+            // The order of these conditionals is IMPORTANT. Failed rows have precedence over Open rows when
+            // determining batch status, and Open rows have precedence over Imported rows. This means all rows
+            // must be marked as Imported for a batch to be marked as Completed, no Failed or Open rows allowed.
+            if (this.hasFailedRow) {
+                return BDI_DataImport_API.BATCH_STATUS_FAILED;
             }
+
+            if (!this.hasOpenRow) {
+                return BDI_DataImport_API.BATCH_STATUS_COMPLETED;
+            }
```

Comment:
How will this hold up if more Data Import Statuses are added in the future?  Or if custom DI Statuses are used in a customer org?  

It seems that this block is returning Failed if the batch has >= 1 failed row, Open if the batch has >= 1 open row, and "Completed" in all other cases.  I know we've been analyzing this scenario and we've probably already answered this, but want to make sure that A) that understanding is correct, and B) that is our expected behavior.  Since, if a Customer decides to overwrite the DI Status of a failed DI row with some custom value (but the row is still a failed row), that Batch's Batch Status might be updated by this code to be "Completed".

Reply:
New statuses would need a priority in the list of statuses. Failed Data Import Statues are most important,  Open Statuses next Imported Statues are the least important. If we had an Intermediate DI/Batch status prioritized between Failed and Open, we'd check for the presence of Intermediate status Data Import rows before checking for Open rows, and return `BDI_DataImport_API.BATCH_STATUS_INTERMEDIATE;` here if one was found. As long as any further Batch Statuses can be prioritized in the list of statuses and the rules around Batch Statuses don't get more complicated, I think this way works.

Reply:
👍🏽